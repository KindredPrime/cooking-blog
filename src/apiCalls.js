import { dummyPosts, dummyComments, dummyUsers } from './dummyData';

const blogPosts = dummyPosts;
const comments = dummyComments;
const users = dummyUsers;

let controller = new AbortController();
let signal = controller.signal;

function abortTasks() {
  controller.abort();
  controller = new AbortController();
  signal = controller.signal;
}

function getAllBlogPosts() {
  return new Promise((resolve, reject) => {
    signal.addEventListener('abort', () => {
      reject('getAllBlogPosts has been aborted');
    });

    resolve(blogPosts)
  });
}

function getBlogPostsByUser(username) {
  return new Promise((resolve, reject) => {
    const userPosts = blogPosts.filter((post) => post.author.username === username);

    signal.addEventListener('abort', () => {
      reject('getBlogPostsByUser has been aborted');
    });

    resolve(userPosts)
  });
}

function getBlogPostById(id) {
  return new Promise((resolve, reject) => {
    const blogPost = blogPosts.find((post) => post.id === parseInt(id));

    if (!blogPost) {
      reject(`There is no blog post with id ${id}`);
    }

    resolve(blogPost);
  });
}

function patchBlogPost(id, content) {
  return new Promise((resolve, reject) => {
    getBlogPostById(id)
      .then((blogPost) => {
        if (blogPost.content === content) {
          resolve(blogPost);
        }

        const newBlogPost = {
          ...blogPost,
          content,
          lastEdited: new Date()
        };

        // Update the post stored in this file
        const postsIndex = blogPosts.findIndex((elem) => elem.id === parseInt(id));
        blogPosts[postsIndex] = newBlogPost;

        resolve(newBlogPost);
      })
      .catch(reject);
  });
}

function getAllComments() {
  return Promise.resolve(comments);
}

function getCommentById(id) {
  return new Promise((resolve, reject) => {
    const comment = comments.find((comment) => comment.id === id);

    if (!comment) {
      reject(`Comment with id ${id} does not exist`);
    }

    resolve(comment);
  });
}

function getCommentsByBlogPost(title) {
  return new Promise((resolve, reject) => {
    const postComments = comments.filter((comment) => comment.blogPost.title === title);

    resolve(postComments);
  });
}

function getCommentsByUser(username) {
  return Promise.resolve(comments.filter((comment) => comment.creator.username === username));
}

function addComment(comment) {
  return new Promise((resolve, reject) => {
    const { content, creator, blogPost } = comment;
    const id = comments.length + 1;

    const newComment = {
      id,
      lastEdited: new Date(),
      content,
      creator,
      blogPost
    }

    comments.push(newComment);

    resolve(newComment);
  });
}

function patchComment(id, content) {
  return new Promise((resolve, reject) => {
    getCommentById(id)
      .then((comment) => {
        if (comment.content === content) {
          resolve(comment);
        }

        const newComment = {
          ...comment,
          content,
          lastEdited: new Date()
        };

        // Update the comment stored in this file
        const commentsIndex = comments.findIndex((elem) => elem.id === id);
        comments[commentsIndex] = newComment;

        resolve(newComment);
      })
      .catch(reject);
  });
}

function deleteComment(id) {
  return new Promise((resolve, reject) => {
    const commentsIndex = comments.findIndex((comment) => comment.id === id);

    if (commentsIndex < 0) {
      reject(`Comment with id ${id} does not exist`);
    }

    const comment = comments[commentsIndex];
    comments[commentsIndex] = {
      ...comment,
      lastEdited: new Date(),
      content: '[Deleted]',
      creator: null,
      deleted: true
    };

    resolve();
  });
}

function getUserById(id) {
  return Promise.resolve(users.find((user) => user.id === parseInt(id)));
}

export {
  abortTasks,
  getAllBlogPosts,
  getBlogPostsByUser,
  getBlogPostById,
  patchBlogPost,
  getAllComments,
  getCommentById,
  getCommentsByBlogPost,
  getCommentsByUser,
  addComment,
  patchComment,
  deleteComment,
  getUserById
};