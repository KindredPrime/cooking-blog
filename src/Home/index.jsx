import { Component } from 'react';
import * as API from '../apiCalls';
import PostsList from '../PostsList/index';

class Home extends Component {
  state = {
    blogPosts: null
  };

  componentDidMount() {
    API.getAllBlogPosts()
      .then((blogPosts) => this.setState({ blogPosts }))
      .catch(console.log);
  }

  componentWillUnmount() {
    API.abortTasks();
  }

  render() {
    const { blogPosts } = this.state;

    return (
      <main className="Home">
        <h1>Welcome to the Cooking Blog!</h1>

        <p>
          Ever wanted to learn more about cooking, or share your experiences with others? Here you can read all kinds of posts from other users: things they've tried, things they want to try, recipes, cooking tips, etc. And if you create an account you can make posts of your own, and leave comments on posts.
        </p>

        {blogPosts && <PostsList initialBlogPosts={blogPosts} />}
      </main>
    );
  }
}

export default Home;