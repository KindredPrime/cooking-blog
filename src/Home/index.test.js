import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import { render, screen, waitFor } from '@testing-library/react';
import Home from './index';
import * as API from '../apiCalls';

describe('Home Component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <BrowserRouter>
        <Home />
      </BrowserRouter>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders an error as expected', async () => {
    const error = 'Test Error';
    const tempOrigAPI = {
      getAllBlogPosts: API.getAllBlogPosts
    };
    API.getAllBlogPosts = () => new Promise((resolve, reject) => {
      throw new Error(error);
    });

    render(
      <BrowserRouter>
        <Home />
      </BrowserRouter>
    );

    await waitFor(() => expect(screen.getByText(error)).toBeInTheDocument());
    expect(document.body).toMatchSnapshot();

    // Reset the API
    API.getAllBlogPosts = tempOrigAPI.getAllBlogPosts;
  });
});