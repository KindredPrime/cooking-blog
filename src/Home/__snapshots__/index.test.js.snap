// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component renders an error as expected 1`] = `
<body>
  <div>
    <main
      class="Home"
    >
      <h1>
        Welcome to the Cooking Blog!
      </h1>
      <p>
        Ever wanted to learn more about cooking, or share your experiences with others? Here you can read all kinds of posts from other users: things they've tried, things they want to try, recipes, cooking tips, etc. And if you create an account you can make posts of your own, and leave comments on posts.
      </p>
      <p
        class="APIError"
      >
        Test Error
      </p>
    </main>
  </div>
</body>
`;
